[클린코드 2기 정현수] 계산기 미션

junghyeonsu opened the pull request:
이번 클린코드 2기 진행하게된 정현수입니다!
기능 구현과 간단한 리팩토링을 마쳐서 우선 리뷰를 요청드립니다.
추후에 테스트 코드 추가와 리팩토링을 지속적으로 할 예정입니다.

### [데모 링크](https://junghyeonsu.github.io/js-calculator/)

### 🎯 기능 요구사항

- [x] 2개의 숫자에 대해 덧셈이 가능하다.
- [x] 2개의 숫자에 대해 뺄셈이 가능하다.
- [x] 2개의 숫자에 대해 곱셈이 가능하다.
- [x] 2개의 숫자에 대해 나눗셈이 가능하다.
- [x] AC(All Clear)버튼을 누르면 0으로 초기화 한다.
- [x] 숫자는 한번에 최대 3자리 수까지 입력 가능하다.
- [x] 계산 결과를 표현할 때 소수점 이하는 버림한다.

### 🚀 추가적으로 구현해야 할 사항들

- [x] 각 테스트들을 파일을 나누어 조금 더 다양한 테스트 작성하기

### ✅ 생각해볼 사항들

- [x] App 파일에 모든 state를 넣고 관리를 하다보니 App이 비대해지는 사항에 대해서

junghyeonsu committed:
chore: 환경 설정

- eslint, prettier 설치 및 설정
- index.js을 index.html에 연결

junghyeonsu committed:
chore: cypress 설치

junghyeonsu committed:
chore: 쓸모없는 테스트 삭제

junghyeonsu committed:
chore: cypress 설정

- cypress-watch-and-reload 라이브러리 설치해서 src 파일들이 변경돼도 자동으로 테스트하도록 설정
- cypress.json 여러가지 설정
  - video, baseUrl, watch 경로

junghyeonsu committed:
test: 계산기 테스트 코드 추가

- cypress를 이용해 테스트 구축

junghyeonsu committed:
chore: selector util 생성

junghyeonsu committed:
feat: App 클래스 생성

- app.js 생성 후 App 클래스를 작성

junghyeonsu committed:
style: 버튼 마우스 hover 커서 효과 추가

junghyeonsu committed:
feat: Digits 컴포넌트 생성

- 이벤트 위임을 이용해 클릭 이벤트 구현

junghyeonsu committed:
feat: Operations 컴포넌트 생성

- Digits 컴포넌트와 비슷하게 이벤트 위임을 이용해 클릭 이벤트 구현

junghyeonsu committed:
feat: Total 컴포넌트 생성

- Digits, Operations 클릭 시 발동하는 이벤트인 recordTotalValue를 구현

junghyeonsu committed:
refactor: Total 컴포넌트 변경

- state를 App으로부터 받고 render하는 방식으로 변경

junghyeonsu committed:
feat: 더하기 연산 구현

junghyeonsu committed:
refactor: 변수명 명확하게 변경

- sum -> evaluationResult

junghyeonsu committed:
feat: 빼기 연산 구현

junghyeonsu committed:
feat: 곱하기 연산 구현

junghyeonsu committed:
feat: 나누기 연산 구현

junghyeonsu committed:
feat: Modifiers 컴포넌트 생성 후 AC 구현

- AC(all clear) 버튼 클릭 시 0으로 리셋

junghyeonsu committed:
test: 테스트 코드 수정

- 조금 더 명확한 버튼을 클릭하도록 get method를 통해서 접근

junghyeonsu committed:
feat: 숫자는 한번에 최대 3자리 수까지 입력 가능 처리

- numberCount state를 추가해서 숫자가 몇 글자 입력됐는지 감시

junghyeonsu committed:
refactor: evaluateDigitsAndOperations 함수 분리

- whatOperatorUseInExpression
- calculateExpressionWithOperator
- calculateReduceWithOperator

세 가지 함수로 모듈화

junghyeonsu committed:
refactor: DOM 상수 객체 생성

junghyeonsu committed:
refactor: Operator, Modifier 상수 객체 생성

junghyeonsu committed:
refactor: Dom, Init state 상수 객체 생성

junghyeonsu committed:
refactor: Message 상수 객체 생성

junghyeonsu committed:
chore: Operation 상수 객체명 변경

- OPERATOR에서 OPERATION으로 변경

junghyeonsu committed:
test: 테스트 코드 정리

- 가독성을 높이기 위해 이벤트들을 함수로 만들어 사용
- 실제 코드에서 사용하는 상수 값들을 사용

junghyeonsu committed:
chore: gitignore 수정

Tanney-102 commented on .gitignore:
```diff
@@ -102,3 +102,10 @@ dist
+
+cypress/fixtures/
+cypress/screenshots/
+cypress/support/
```
L106 ~ L111은 git에 올라오면 안되는 내용들인가요?
이미 올라와 있는 것 같긴하네요 ㅎㅎ

Tanney-102 commented on cypress/integration/calculator.spec.js:
```diff
@@ -0,0 +1,75 @@
+  cy.on('window:alert', text => {
+    expect(text).to.contains(message);
+  });
+const THE_RESULT_SHOULD_BE = result => cy.get(DOM.total).should('have.text', result);
```
이런 함수들을 cypress commands로 선언하는 게 어떨까요?
https://docs.cypress.io/api/cypress-api/custom-commands

Tanney-102 replied:
보통 상수를 표현할때 upper snake case를 사용하는데 함수를 이렇게 표현하신 의도가 있으실까요?

Tanney-102 commented on package.json:
```diff
@@ -0,0 +1,32 @@
+    "prettier": "^2.5.1"
+  },
+  "dependencies": {
+    "cypress-watch-and-reload": "^1.7.4"
```
devDenpendencies로 가야할 것 같아요!

junghyeonsu replied:
<img width="551" alt="스크린샷 2022-03-13 오후 3 17 38" src="https://user-images.githubusercontent.com/54893898/158047699-6118a528-82ed-4462-8fee-5ef543b0fc22.png">

<img width="506" alt="스크린샷 2022-03-13 오후 3 19 03" src="https://user-images.githubusercontent.com/54893898/158047706-4e694638-24dd-4d63-bf5a-917d14107f30.png">

저도 `devDenpendencies`가 맞다고 생각했는데, `devDenpendencies`로 옮기니까 위와 같이 `dependencies`에 있어야 한다고 뜨더라구요! 공식문서에도 `npm install --save-dev` 가 아닌 그냥 `npm install`로 설명되어 있습니다!

- [공식문서](https://github.com/bahmutov/cypress-watch-and-reload)

Tanney-102 commented on src/js/app.js:
```diff
@@ -0,0 +1,103 @@
+    this.state = {
+      currentTotal: INIT_STATE.currentTotal,
+      numberCount: INIT_STATE.numberCount,
+    };
```
```suggestion
    this.state = { ...INIT_STATE };
```
이렇게 쓰지 않으시는 이유는 INIT_STATE가 this.state와 모양이 다를 수도 있기 때문일까요?

Tanney-102 commented on src/js/app.js:
```diff
@@ -0,0 +1,103 @@
+      this.recordOperation(clickedValue);
+    } else {
+      this.recordDigit(clickedValue);
+    }
```
App에서 이 모든 내용을 다 알고 있어야할까요?
이 부분을 각 자식 컴포넌트로 옮길 수 있는 방법이 있을지 고민해보시면 좋겠습니다!

Tanney-102 commented on src/js/app.js:
```diff
@@ -0,0 +1,103 @@
+  }
+
+  recordOperation(operation) {
+    if (this.state.currentTotal === INIT_STATE.currentTotal) {
```
함수나 변수를 사용해 조건문의 의도를 명확하게 밝히면 어떨까요?
에를 들어
`const isDirty = this.state.currentTotal !== INIT_STATE.currentTatal`
이런식으로요!
지금은 어떤 이유로 분기가 되는 것인지 직관적이지 않습니다!

L51에서도 동일한 조건문이 사용되니 메서드로 만들어도 좋겠네요!

Tanney-102 commented on src/js/app.js:
```diff
@@ -0,0 +1,103 @@
+        numberCount: this.state.numberCount + 1,
+      });
+    }
+  }
```
if문을 어떻게 쓰냐는 개인적인 스타일 차이일 수 있긴 하지만 이 부분은 조금 정리되면 좋겠네요~

현재 이 함수 안에는 조건에 따라 3가지 동작이 수행됩니다.
1. 기존 currentTotal에 새 digit을 연결
2. currentTotal을 제거하고 새 digit을 등록
3. alert을 통한 예외처리

그런데 코드를 읽는 입장에서 어떤 상황에 어떤 일이 벌어지는 건지 한 눈에 파악하기가 어렵습니다. 조건문이 직관적이지 못한 이유도 있겠지만 맥락없이 기능만을 위한 코드가 나열되어 있는 것 같아요!
같은 단어를 쓰더라도 어떻게 배치하느냐에 따라 뉘앙스가 달라지는 것처럼 if문 하나도 어떻게 쓰냐에 따라 다른 뉘앙스를 풍길 수 있습니다. 저는 개인적으로 예외 상황을 처리할 때는 early return을 통해 if문을 따로 분리합니다. 그리고 '같은 일을 컨텐츠만 다르게 실행'하는 경우에 주로 if~else나 switch같은 분기를 사용합니다! 위 코드를 이러한 방식으로 작성하면 다음과 같이 되겠죠!
```js
  recordDigit(digit) {
     if (this.state.numberCount >= 3) {
       alert(MESSAGE.numberCanBeEnteredUpToThreeDigitsAtOnce);
       return;
     } 
  
    if (this.state.currentTotal === INIT_STATE.currentTotal) {
      this.setState({ currentTotal: digit, numberCount: this.state.numberCount + 1 });
    } else {
      this.setState({
        currentTotal: this.state.currentTotal + digit,
        numberCount: this.state.numberCount + 1,
      });
    }
  }
```

이렇게 분리하고 나면 때에 따라 더 간단히 작성할 수 있는 길이 보이기도 합니다!
예를 들어 다음과 같이요!
```js
  recordDigit(digit) {
     if (this.state.numberCount >= 3) {
       alert(MESSAGE.numberCanBeEnteredUpToThreeDigitsAtOnce);
       return;
     } 
  
    const isDirty = this.state.currentTotal !== INIT_STATE.currentTotal
    const prevTotal = isDirty ? '' : this.state.currentTotal
    this.setState({ currentTotal: prevTotal + digit, numberCount: this.state.numberCount + 1   }
}
```

Tanney-102 commented on src/js/app.js:
```diff
@@ -0,0 +1,103 @@
+    }
+  }
+
+  evaluateDigitsAndOperations(expression) {
+    const operation = this.whatOperatorUseInExpression(expression);
+    this.calculateExpressionWithOperator(expression, operation);
+  }
```
this.calculateExpressionWithOperator 안에서 whatOperatorUseInExpression을 호출하면 굳이 이 메서드를 만들 필요가 없지 않았을까요?

Tanney-102 commented on src/js/app.js:
```diff
@@ -0,0 +1,103 @@
+    });
+  }
+
+  calculateReduceWithOperator(acc, operation, cur) {
+    switch (operation) {
+      case OPERATION.plus:
+        return Number(acc) + Number(cur);
+      case OPERATION.minus:
+        return Number(acc) - Number(cur);
+      case OPERATION.multiple:
+        return Number(acc) * Number(cur);
+      case OPERATION.division:
+        return Math.floor(Number(acc) / Number(cur));
+      default:
+        return false;
+    }
+  }
+}
```
이 클래스 안에 멤버로 있을 필요는 없는 함수인 것 같습니다.

그리고 
```js
  executor = {
      [OPERATION.plus]: (left, right) => left + right,
      [OPERATION.minus]: (left, right) => left - right,
      [OPERATION.multiple]: (left, right) => left * right,
      [OPERATION.division]: (left, right) => Math.floor(left / right),
  }
  executor[operator](Number(acc), Number(cur))
```
요렇게 분기하는 방법도 있으니 참고해주세요~

Tanney-102 commented on src/js/components/Total.js:
```diff
@@ -0,0 +1,17 @@
+    this.state = state;
+  }
+
+  setState(nextState) {
+    this.state = nextState;
+    this.render();
+  }
+
+  render() {
+    this.$target.innerHTML = this.state;
+  }
```
그냥 render 메서드에서 totalValue를 인자로 받아 dom에 넣어주면 될 것 같은데 setState를 따로 두는 이유가 있을까요?

junghyeonsu replied:
https://prgms.tistory.com/53

<img width="825" alt="스크린샷 2022-03-13 오후 5 26 02" src="https://user-images.githubusercontent.com/54893898/158051436-d6b1c6ff-b41a-4223-8329-03d0181cda3f.png">

제가 예전에 프로그래머스 과제인 고양이 사진첩을 구현하고, 해설 블로그를 봤는데
자식 컴포넌트들은 자신의 state를 기준으로 렌더링을 진행하기 때문에 별도의 인자를 받지 않는다고 하더라구요!

이게 정답은 아니지만, 부모 컴포넌트에서 setState를 진행을 하면 그 state를 기준으로 render가 진행이 된다. 라고 생각을 하니까 조금 더 깔끔해서 이 방식을 사용하고 있습니다!

Tanney-102 reviewed (changes requested):
안녕하세요 현수님!
계산기 미션 리뷰어 오태은입니다 ㅎㅎ

작성하신 코드 잘봤습니다! 각 역할을 나눠 컴포넌트를 구현하려고 노력하신게 인상깊었어요!
다만 컴포넌트를 여러개 둔 것에 비해 역할이 그리 잘 나눠져있는 것 같진 않네요~
App 컴포넌트 하위에 몇몇 자식 컴포넌트들이 있긴한데 사실상 사용자의 입력을 App에 전달하는 것 말고는 별 역할을 하고 있지 않습니다. 관련하여 코멘트 남겼으니 한 번 고민해보시면 좋을 것 같아요!
그리고 컴포넌트 안에서(app, total) setState 메서드를 구현해 상태를 dom에 반영하려고 하신 것 같아요!
그런데 꼭 setState가 필요했을지 고민해보셨으면 좋겠어요. 제가 보기에는 필요해서라기보다 구현을 위한 구현 같습니다. 리액트 같은 프레임워크에서 왜 setState를 구현하는지도 생각해보시면 좋겠네요~

분기문에 관해서도 몇몇 코멘트 남겨놨습니다! 확인하시고 고민해보시면 좋겠습니다!

Tanney-102 commented:
아 그리고! 추가로 버그가 있어요!
`9X8+3` 과 같은 방식으로 여러 연산자가 섞여있으면 NaN을 뱉습니다!
연산자가 하나만 들어온다고 가정하시고 구현한 것 같은데 수정되면 좋겠네요~

Tanney-102 commented:
<img width="426" alt="스크린샷 2022-03-12 오후 7 38 30" src="https://user-images.githubusercontent.com/57767891/158014734-edde761f-642a-4c03-94d3-e11d595bc28a.png">
연산자가 여러개 들어가는 이슈도 있네요~

junghyeonsu committed:
test: cypress command 생성 및 테스트 수정

- cypress commands를 생성해서 함수를 만듦
- gitignore에서 cypress/support 제거

junghyeonsu committed:
chore: Message 객체명 변경

- pleaseEnterNumberBeforeOperator가 아닌 pleaseEnterNumberBeforeOperation으로 변경

junghyeonsu committed:
test: 테스트 케이스 추가

- 연산자 앞에는 숫자가 있어야 한다.
  - 아무 입력 없이 연산자를 클릭하면 alert를 띄움
- 연산자를 연속으로 누를 수 없다.
  - '2++' 이런식의 식은 안되도록 테스트 케이스 생성

junghyeonsu committed:
feat: 연산자 연속으로 입력 불가능하게 구현

- lastClickedButton state를 선언해서 마지막 입력이 연산자면 alert로 메세지를 호출

junghyeonsu committed:
refactor: 조건문을 함수로 변경

- this.state.currentTotal === INIT_STATE.currentTotal를 isClean() 이라는 함수로 변경

junghyeonsu committed:
refactor: onClickDigit 생성

- onClickButtons 함수에서 onClickDigit만 따로 빼서 digit 클래스에 전달

junghyeonsu committed:
refactor: onClickButtons에서 onClickModifier 함수 추출

junghyeonsu committed:
refactor: executor, validator util 객체 생성

- App 클래스에서 계산하지 않고, executor util obj를 생성해서 계산 로직을 분리

junghyeonsu committed:
chore: 사용하지 않는 recordDigit 함수 삭제

- onClickDigit로 변경됨

junghyeonsu committed:
refactor: operation 추출기 생성

- operation을 추출하는 함수를 util 함수로 생각해서 옮김

junghyeonsu committed:
test: 테스트 코드 작성

- "=" 클릭 시 맨 마지막 클릭한 버튼은 숫자여야 한다.

junghyeonsu committed:
refactor: onClickOperation 정리

- validateWhenClickOperation 함수를 생성해 검증하는 함수를 따로 생성
- 쓸모없는 함수들 (calculateExpression, recordOperation)들을 삭제하고 변수 명들을 조금 더 명확하게 변경

junghyeonsu commented on cypress/support/commands.js:
```diff
@@ -0,0 +1,21 @@
+import { DOM } from '../../src/js/constants.js';
+
+Cypress.Commands.add('clickDigit', digit =>
+  cy.get(`${DOM.digits} > ${DOM.digit}`).contains(digit).click(),
+);
+
+Cypress.Commands.add('clickOperation', operation =>
+  cy.get(`${DOM.operations} > ${DOM.operation}`).contains(operation).click(),
+);
+
+Cypress.Commands.add('clickModifier', modifier =>
+  cy.get(`${DOM.modifiers} > ${DOM.modifier}`).contains(modifier).click(),
+);
+
+Cypress.Commands.add('checkAlertMessage', message =>
+  cy.on('window:alert', text => {
+    expect(text).to.contains(message);
+  }),
+);
+
+Cypress.Commands.add('totalShouldBe', result => cy.get(DOM.total).should('have.text', result));
```
피드백 주신 부분들을 commands로 옮겼습니다!

junghyeonsu commented on src/js/app.js:
```diff
@@ -0,0 +1,97 @@
+      return;
+    }
+
+    if (this.$total.isClean()) {
```
분기문을 함수로 추출해서 의도를 파악하기 쉽게 했습니다!

junghyeonsu commented on src/js/app.js:
```diff
@@ -0,0 +1,97 @@
+    }
+  }
+
+  validateWhenClickOperation(operation) {
+    if (operation === OPERATION.equal && validator.isOperation(this.state.lastClickedButton)) {
+      alert(MESSAGE.lastCharacterMustBeNumber);
+      return false;
+    }
+    if (this.$total.isClean()) {
+      alert(MESSAGE.pleaseEnterNumberBeforeOperation);
+      return false;
+    }
+    if (validator.isOperation(this.state.lastClickedButton)) {
+      alert(MESSAGE.operationCannotBeEnteredConsecutively);
+      return false;
+    }
+    return true;
+  }
```
`Operation` 관련해서는 예외 상황이 많아서 따로 함수로 추출해서 검증했습니다!

junghyeonsu reviewed (commented):
정말 상세한 리뷰 감사합니다!
제가 생각지도 못했던 부분들이 많아서 신기하고 재미있었어요!
제가 리뷰해주신 부분들 혼자 생각해보면서 질문사항이 몇 가지가 생겼습니다!

지금은 여러 연산자가 들어갔을 때는 구현하지 못했습니다. 리뷰 해 주신 부분들을 잘 생각하면서 리팩토링 한 것을 또 피드백 받고 싶어서 이렇게 리뷰 요청 드렸습니다!

### 추가적으로 구현해야 하는 것

- [ ]  여러 연산자 들어갔을 때 구현하기

## 질문사항

### 1. 자식 컴포넌트 Dom 변수 위치에 대해서

![image](https://user-images.githubusercontent.com/54893898/158051508-e7f50f60-fdc6-4723-abe4-79d30fae4338.png)


현재는 app.js 코드에서 자식 컴포넌트들을 위와 같이 선언하고 사용하고 있는데요

자식 컴포넌트들에 첫번째 인자로 DOM 객체를 넘겨주는 방식으로 하고 있는데

이 Dom객체를 자식 컴포넌트에서 불러서 쓰는게 나을까요? 아니면 현재처럼 app.js에서 넘겨주는게 깔끔할까요?.. 감이 오질 않습니다! 멘토님은 어떻게 하실 것 같나요?

### 2. 컴포넌트에 대해서

현재는 모든 state를 app.js에 생성해서 사용하고, state들이 변경시키고 싶을 때 setState를 불러서 자식 컴포넌트들을 해당 state를 기준으로 rendering하는 방식으로 사용하고 있습니다.

지금 컴포넌트에서 state를 사용하는 곳이 total 컴포넌트 밖에 없어서 리렌더링을 위해서 컴포넌트를 나눴다기 보다는 이벤트리스너마다 컴포넌트를 나누어서 진행을 했다?.. 라는 느낌이 더 강한데요 혹시 멘토님은 이런 상황에서는 어떤 방식으로 컴포넌트를 나누는게 낫다고 생각하시나요!?

### 3.  window.Function 사용

```jsx
const evaluatedExperssion = window.Function(`return ${expression.replace('X', '*')}`)();
console.log(evaluatedExperssion);
```

위와 같이 window.Function을 사용하면, string을 계산하도록 되어있는데 사용하지 않는 이유는 성능상의 문제나 다른 문제가 있을까요?

junghyeonsu committed:
test: 더하기 연산 테스트 파일 추출

- 3개의 숫자에 대해 덧셈이 가능한 지 테스트
- 4개의 숫자에 대해 덧셈이 가능한 지 테스트

junghyeonsu committed:
test: beforeEach를 support/commands 파일로 이동

- 모든 테스트 파일에 선언할 필요없이 support/commands 파일에만 선언해주면 됨

junghyeonsu committed:
test: 나누기 연산 테스트 파일 생성

junghyeonsu committed:
test: 예외 처리 테스트 파일 생성

junghyeonsu committed:
test: 빼기 연산 테스트 파일 생성

junghyeonsu committed:
test: 수정자 테스트 파일 생성

junghyeonsu committed:
test: 곱하기 연산 테스트 파일 생성

junghyeonsu committed:
test: 복잡한 연산 테스트 작성

- 더하기 곱셈, 곱셈 빼기와 같이 우선순위가 다른 연산자들이 섞였을 때의 테스트 생성

junghyeonsu committed:
chore: 복잡한 연산 테스트 삭제

junghyeonsu committed:
Revert "chore: 복잡한 연산 테스트 삭제"

This reverts commit dae667e8014321c263360953e6dde20025560999.

junghyeonsu commented:
복잡한 연산들 예를들어 `2 + 2 x 2 - 2` 와 같은 연산을 처리해보려고 시도를 했는데,,, 잘 안되네요ㅜㅜㅜㅜ
중위표현식을 후위표현식으로 바꾸고, 후위 표현식을 계산해서 처리하는 방식도 시도해보고
받은 expression을 연산자 기준으로 나눠서 계산하는 방식도 해보았는데 잘 되지 않았습니다...
시간이 남는다면 조금 더 구현을 해보겠습니다!

Tanney-102 reviewed (changes requested):
현수님! 원래도 구현을 잘 해주셨는데 코드가 더 깔끔해졌네요!
테스트 코드도 성공 케이스부터 예외까지 잘 작성해주셨네요~
complexOperation.spec 에 해당하는 내용은 아직 구현하시기 전으로 인지했는데 맞을까요?
이 부분은 우선 넘어가도 좋을 것 같습니다! 미션의 취지를 넘어서는 영역인 것 같아서요~
확인하시면 머지하도록 하겠습니다!

질문에 대한 답변입니다.
1. 저는 개인적으로 지금 구현하신 방법을 선호하긴 합니다. 컴포넌트가 특정 dom을 알고 있는 것보다 인자로 받는 것이 재사용에 유리하니까요~ 
사실 지금도 Digits, Modifiers, Operations가 하는 일이 다 같아서 DOM selector만 추가로 인자로 주입받으면 하나의 컴포넌트를 재사용해 모두 대체 가능합니다. 그리고 더 나아가 class로 구현할 필요도 없이 함수 하나로 다 가능합니다.

2. 우선 1번의 답변을 참고해주시구요~ 
사실 계산기는 하나의 '컴포넌트'로 충분한 어플리케이션이라 생각합니다.
어떤 컴포넌트로 나눌지 고민하기 이전에 컴포넌트가 무엇이고, 어떤 역할을 하기 바라는지 명확해야할 것 같습니다!

3. 지금 상황에선 크게 문제 없어보입니다.

junghyeonsu committed:
feat: 복잡한 연산 구현

- window.Function을 이용해서 구현

junghyeonsu commented:
@Tanney-102 내용 확인했습니다!

다음 미션 때부터는 컴포넌트를 구분하는거에 대해서 조금 더 깊게 생각을 해봐야겠군요!
컴포넌트로 나누는 이유가, 어떤 역할을 해야하고, 꼭 필요한 분리인가?에 대해서 생각해보겠습니다!

복잡한 연산 테스트로직은 `window.Function` 이용해서 구현을 해놓았구요!
테스트코드는 현재 전부 통과를 합니다!

꼼꼼한 리뷰 정말 감사했습니다!~

Tanney-102 reviewed (approved):
미션 진행하느라 고생많으셨습니다~
